prov/verbs/src/verbs_info.c

```diff
@@ -751,9 +752,24 @@ static int vrb_have_device(void)
 	for (i = 0; devs[i]; i++) {
 		verbs = ibv_open_device(devs[i]);
 		if (verbs) {
+			ret = ibv_query_device(verbs, &attr);
 			ibv_close_device(verbs);
-			ret = 1;
-			break;
+			if (ret)
+				break;
+			/*
+			 * EFA device is a RDMA device, but it does not support
+			 * the rdma_cm required by verbs provider. To avoid
+			 * failure in the case of an efa device selecting verbs,
+			 * check the device vendor id. For Amazon devices
+			 * vendor id is 0x1d0f. If all devices are Amazon
+			 * devices then return 0
+			 */
+			if (attr.vendor_id == 0x1d0f) {
```

Comment:
Isn't EFA usable for UD QPs?  The location of this check doesn't look correct.  If EFA were ever in a system with another verbs device, rdma_get_devices() would still be called and open it.  (Yes, I know the environment is controlled, but this change seems to have that assumption built in for this to work.)

Also, it would be better if 1d0f were some constant, to make this a little more readable.  E.g.:
```
ibv_close_device(verbs);
if (ret == 0 && attr.vendor_id != AWS_VENDOR_ID) {
    ret = 1;
    break;
}
```

The problem still isn't clear to me.  Does EFA have a restriction on how many times the device can be opened?  If the RDMA CM opens EFA, but really can't use it at all, should there also be a patch against the librdmacm to close the device and remove it from its list?

Reply:
> Isn't EFA usable for UD QPs? 

It is, EFA is usable via `IBV_QPT_UD` and `IBV_QPT_DRIVER` (for the SRD-capable endpoint). We have not tested the verbs provider with EFA's UD QP, since it is already made available to applications via the EFA provider.

> Does EFA have a restriction on how many times the device can be opened?

Yes, we have a PD limit of 128. When fully subscribing an EC2 instance with 72 or 96 ranks, this leak holds on to an additional PD for every process and causes issues. Not an rdmacm expert, but it seems odd to me that the resources allocated using one library is expected to be freed by another. 

Given the `rdma_get/free_devices()` semantics, it might make sense to handle this in librdmacm like you say, but there's still an assumption here that the verbs provider is going to serve any device in the RDMA subsystem. It might be good to filter out devices that do not support all the QP types that verbs needs or possibly have an explicit blacklist for cases like this where we have a clear preference for which provider should serve a device. @galpress, thoughts?

Another short-term fix would be to change the provider priorities. Registering EFA before verbs would prevent this issue.

Reply:
The librdmacm does opens the devices and keeps them open.  It does not expect some other library or app to close them.  In fact, doing so breaks the librdmacm.  rdma_get_devices() is only one of the calls that opens the devices.  Other calls can open them as well.  That's why rdma_free_devices() does not close them.

The librdmacm supports UD QPs, so EFA would be supported.

If verbs devices are not available on AWS systems, why not disable the provider?

I don't think changing the provider priorities fixes this.  The librdmacm will still open the device.

Reply:
I think my proposed change above would result in the verbs provider doing nothing unless there's a non-EFA verbs device found in the system.  I'm okay with that.  We just need to comment that we don't expect EFA and other verbs devices to be in the same system, and EFA is handled by the EFA provider.

Reply:
> If verbs devices are not available on AWS systems, why not disable the provider?

This is what we have been doing with the [EFA installer builds](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa-start.html#efa-start-enable). Trying to find a path where a default configuration will work just as well. Intel MPI has both providers built as DSOs, so can't think of a better solution when using its internal version of OFI that discovers both.

> I don't think changing the provider priorities fixes this. The librdmacm will still open the device.

You are right, this is a race to which INI function opens devices across all the processes. Ignore this suggestion then.